embassy/src/util/select_all.rs

```diff
@@ -0,0 +1,56 @@
+#![allow(dead_code)]
+
+use core::future::Future;
+use core::pin::Pin;
+use core::task::{Context, Poll};
+
+use futures::future::FutureExt;
+
+/// Future for the [`select_all`] function.
+#[derive(Debug)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct SelectAll<Fut, const N: usize> {
+    inner: [Fut; N],
+}
+
+impl<Fut: Unpin, const N: usize> Unpin for SelectAll<Fut, N> {}
+
+/// Creates a new future which will select over a list of futures.
+///
+/// The returned future will wait for any future within `iter` to be ready. Upon
+/// completion the item resolved will be returned, along with the index of the
+/// future that was ready.
+///
+/// # Panics
+///
+/// This function will panic if the array specified contains no items.
+pub fn select_all<Fut: Future + Unpin, const N: usize>(arr: [Fut; N]) -> Option<SelectAll<Fut, N>> {
```

Comment:
You can make this not require Unpin, which makes it much easier to use (no `pin_mut!()` required!). 

(`.into_inner()` requires Unpin, maybe you can make just that one require Unpin.

Reply:
You're right, that's much better. I think the futures version requires `Unpin` because it uses a `Vec` so the futures get moved around.

I'd appreciate a review of the safety in the `poll()` implementation, now that it has to use unsafe.